### PR TITLE
MDEV-19210: use environment file in systemd units

### PR DESCRIPTION
**Pull request by eworm-de**
- [x] *The Jira issue number for this PR is: MDEV-10210*

## Description
We used to run `systemctl set-environment` to pass `_WSREP_START_POSITION`. This is bad because:

* it clutter `systemd`'s environment (yes, pid 1)
* it requires root privileges
* options (like `LimitNOFILE=`) are not applied

Let's just create an environment file in `ExecStartPre=`, that is read before `ExecStart=` kicks in. We have `_WSREP_START_POSITION` around for the main process without any downsides.

## How can this PR be tested?

Well, good question... This should not add new regressions and continue as expected.

Special care is to be taken for Debian, which adds extra commands to the systemd service file.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/11.0/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

**Comment by illuusio** [comment]
Ok I like nice adjustements with `+` sigh for Debian but then I just like to ask why should `_WSREP_START_POSITION` be on systemd file? As most of the people does not need it (like me who don't use galera but likes to have normal replication)? It would be easier to manage to have this systemd privilege thing in another PR ja wsrep/replication in another.
Should this be worked to work on RPM side too?

**Comment by eworm-de** [comment]
> Ok I like nice adjustements with `+` sigh for Debian but then I just like to ask why should `_WSREP_START_POSITION` be on systemd file?

Currently the service file handles single server and Galera cluster. I do not change that, I just change the way to pass environment variables.

> As most of the people does not need it (like me who don't use galera but likes to have normal replication)? It would be easier to manage to have this systemd privilege thing in another PR ja wsrep/replication in another.

This is not about adding functionality, this is just about changing an implementation detail.

> Should this be worked to work on RPM side too?

Only if there is any mechanism that adds more `ExecStartPre=` or `ExecStartPost=` for RPM packages that needs root privileges. I do not think there is anything like that.

**Comment by LinuxJedi** [comment]
I pushed this to a bb branch to run through old buildbot to make sure it doesn't fail, results: https://buildbot.mariadb.net/buildbot/grid?category=main&branch=bb-10.4-linuxjedi-MDEV-19210

**Comment by LinuxJedi** [comment]
Requesting a review from @sysprg for this.

**Comment by eworm-de** [comment]
Thanks a lot! Looks promising I think.

**Comment by eworm-de** [comment]
Any news on this?

I could not find failures related to service startup. So should be ok now?

**Comment by eworm-de** [comment]
Ping... Any news or is this going to stall mode for years again?

**Comment by LinuxJedi** [comment]
Sorry @eworm-de, I have on a combination of vacation and working on some other high-priority things. I'm starting to catch up.

Unfortunately I cannot merge this without approval from @sysprg, due to the ownership of the Galera section of the codebase. This message will have pinged him again.

**Comment by grooverdan** [comment]
pushed to [bb-10.4-MDEV-19210-environment-file-pkgtest](https://buildbot.mariadb.org/#/grid?branch=bb-10.4-MDEV-19210-environment-file-pkgtest) to create files for testing. A 11.3/11.4 might be a more realistic target.

**Comment by eworm-de** [comment]
Well, this issue is that old... I think I started with 10.2. 😜 That's why it is stuck at 10.4 right now.

But I am free to anything else... Let me know and I will change the target. On the other side - I would like to see it in a GA release any time soon.

**Comment by grooverdan** [comment]
@eworm-de hi,

Started testing on https://github.com/MariaDB/mariadb.org-tools/blob/master/daniel/MDEV-19210 - so an alma linux

I just want to make sure I've covered main test cases:
* galera_new_cluster
* nodes can join
* add data
* disconnect node
* add mode data
* rejoin node performs SST and pulls recovery position

What else do I need to test? (I'll do the same of a few more distro types).

As you may not have expected on Alma/RHEL - datadir is `@mysqlunixdir@`. So do you think that needs a modification?
From https://github.com/devexp-db/mysql-selinux/blob/master/mysql.fc - `/var/run/mariadb` might be safer.


Aside: have you thought about https://systemd.io/FILE_DESCRIPTOR_STORE/ - would require much more server implementation.

File: support-files/mariadb.service.in
Comment (grooverdan) [inline support-files/mariadb.service.in:72]: `%N` would refer to a templated systemd service which is `mariadb@.service.in`. Would it make sense to include it there. I'm thinking remove it here and put it in `mariadb@.service.in`.

If that's true then, the `wsrep-start-position` would need a `-%N` suffix too to avoid conflicts in the same template file.

File: scripts/galera_new_cluster.sh
Comment (grooverdan) [inline scripts/galera_new_cluster.sh:24]: was the drop in execute permissions on this file intentional? Handled in CMakeList.txt, but asking anyway.

File: support-files/mariadb.service.in
Comment (grooverdan) [inline support-files/mariadb.service.in:87]: `wsrep-start-postition-%N` in `mariadb@.service.in` for consistency.

File: support-files/mariadb.service.in
Comment (eworm-de) [reply support-files/mariadb.service.in:72]: `%N` is a specifier by systemd and evaluates to *Full unit name*, see [systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html) and search for it.

When I started this (back in 2018?) Galera cluster was supported from instantiated services. Probably we can drop this now...

File: scripts/galera_new_cluster.sh
Comment (eworm-de) [reply scripts/galera_new_cluster.sh:24]: What execute permissions? Can you elaborate?

Comment (grooverdan) [reply scripts/galera_new_cluster.sh:24]: image

File: support-files/mariadb.service.in
Comment (eworm-de) [reply support-files/mariadb.service.in]: Dropped specifier in rebase.

File: support-files/mariadb.service.in
Comment (eworm-de) [reply support-files/mariadb.service.in:87]: The instantiated service `mariadb@.service` does not support Galera cluster. Support was dropped in 91f1694836b65afd7fe2f4a669106e51ee47612c.

File: scripts/galera_new_cluster.sh
Comment (eworm-de) [reply scripts/galera_new_cluster.sh]: Ah, got it. Good question... Probably I've wanted to rename it to `galera_new_cluster.sh.in` as it is a template now...

**Comment by eworm-de** [comment]
Rebased to rename `galera_new_cluster.sh` to `galera_new_cluster.sh.in` as it is a template now.

**Comment by eworm-de** [comment]
> @eworm-de hi,
> 
> Started testing on https://github.com/MariaDB/mariadb.org-tools/blob/master/daniel/MDEV-19210 - so an alma linux
> 
> I just want to make sure I've covered main test cases:
> 
>     * galera_new_cluster
> 
>     * nodes can join
> 
>     * add data
> 
>     * disconnect node
> 
>     * add mode data
> 
>     * rejoin node performs SST and pulls recovery position
>
> What else do I need to test? (I'll do the same of a few more distro types).

Sounds good.

This PR also effects the regular (non-Galera) use, as it drops `PermissionsStartOnly=true` - that broke things for Debian in an earlier revision (Debian using `ExecStartPre` & `ExecStartPost` was not covered). But it is fixed now and should be covered in your tests anyway. If you want to make extra sure add a regular standalone start.

> As you may not have expected on Alma/RHEL - datadir is `@mysqlunixdir@`. So do you think that needs a modification? From https://github.com/devexp-db/mysql-selinux/blob/master/mysql.fc - `/var/run/mariadb` might be safer.

Uh, what?! Where goes the socket then?

Anyway, should not be a problem I think... The daemon should not read the file for anything.

> Aside: have you thought about https://systemd.io/FILE_DESCRIPTOR_STORE/ - would require much more server implementation.

That is something completely different. I do not see how this could make sense here.

This feature is to store file **descriptors** in the manager. So you could make `mariadbd` use that to store incoming socket connections, and retrieve them later - from another process after restart. That way you could install an upgrade from mariadb 11.2.2 to mariadb 11.2.3, restart the daemon - and the clients would continue to use the old connections without interruption as these are still valid.

**Comment by eworm-de** [comment]
> > > As you may not have expected on Alma/RHEL - datadir is `@mysqlunixdir@`. So do you think that needs a modification? From https://github.com/devexp-db/mysql-selinux/blob/master/mysql.fc - `/var/run/mariadb` might be safer.
> > 
> > 
> > Uh, what?! Where goes the socket then?
> 
> ```
> MariaDB [test]>  \s
> --------------
> mariadb from 11.4.0-MariaDB, client 15.2 for Linux (x86_64) using  EditLine wrapper
> ...
> UNIX socket:		/var/lib/mysql/mysql.sock
> ```

Does SST transfer that socket to other nodes then?

> > Anyway, should not be a problem I think... The daemon should not read the file for anything.
> 
> And if a SST copies these files to another node, will the systemd service there use them?

That should not happen. The files are removed immediately after they are no longer used: `wsrep-start-position` by `ExecStartPost=` in `mariadb.service`, `wsrep-new-cluster` in `galera_new_cluster`. When a node becomes donor it is clean.

**Comment by eworm-de** [comment]
> Was thinking the recovery position in a memfd while the mariadbd goes offline for SST. I need to look at the sequence of events and data. But probably won't have time before 11.4, so I should just keep on track.

Note that you can not store files there, just open file **descriptors** or handles. So the file would have to exist in filesystem anyway.

**Comment by grooverdan** [comment]
> Rebased to rename `galera_new_cluster.sh` to `galera_new_cluster.sh.in` as it is a template now.

not sure why, .sh files where templated just fine. Reverted by pushing.

**Comment by grooverdan** [comment]
Debian post start/upgrade functionality is working.
Added https://github.com/MariaDB/mariadb.org-tools/tree/master/daniel/MDEV-19210/deb, but something odd in its network config. Still testing.

**Comment by eworm-de** [comment]
The next and last version from `11.4` branch will be released in a few days? Guess this will not make it then?

Should I rebase on `11.5`?

**Comment by LinuxJedi** [comment]
> The next and last version from `11.4` branch will be released in a few days? Guess this will not make it then?
> 
> Should I rebase on `11.5`?

Hi @eworm-de, it will need rebasing against 11.5 or 11.6 (depending on which one it will land it). We can do this after review if it helps.

I'll take a look at the more recent changes shortly and see if I can push this along.

**Comment by eworm-de** [comment]
Rebased and switched to `11.5`...

**Review by LinuxJedi — APPROVED** [review]
Looks like everything is addressed and it looks like a nice improvement to me.

**Comment by LinuxJedi** [comment]
Whilst I have approved this, the MTR is showing Galera failures in old Buildbot for the branch @grooverdan created for it (https://buildbot.mariadb.net/buildbot/grid?category=main&category=package&branch=bb-11.4-pr2726-MDEV-19210-environment-file-pkgtest). I don't know whether this is related or not. Someone with Galera knowledge would need to check.

**Comment by eworm-de** [comment]
Any news on this? Perhaps pushing the current branch to the buildbot makes sense...

**Comment by LinuxJedi** [comment]
> Any news on this? Perhaps pushing the current branch to the buildbot makes sense...

Hi @eworm-de, 11.6 is open, so we will work towards getting it in there. I'll try to sort out a second reviewer for you.

**Comment by eworm-de** [comment]
Oh, I had just rebased this on `11.5` a month ago... Should I switch and rebase on `11.6` now?

**Review by grooverdan — APPROVED** [review]

**Comment by LinuxJedi** [comment]
> Oh, I had just rebased this on `11.5` a month ago... Should I switch and rebase on `11.6` now?

Sorry, I was wrong, there isn't an 11.6 just yet, but it should exist any day now. It will need rebasing to that once it exists.

**Comment by eworm-de** [comment]
Rebased on `11.6`. Any chance for some progress now?

**Comment by LinuxJedi** [comment]
> Rebased on `11.6`. Any chance for some progress now?

There is a test failing (even after retriggers) which I don't think is related to your work, but I physically cannot merge it until it passes. We will look into this.

**Comment by LinuxJedi** [comment]
All good, thanks again!

**Comment by grooverdan** [comment]
:partying_face: :balloon: . Thank you.

**Pull request merged**
